Sync bug fixes from v2.6

guilhermeblanco opened the pull request.

Tobion committed:
Fix applying collation on foreign key columns

Commit:
Fix handling entities with post generated IDs as FK

This prevents a throw in UnitOfWork#addToIdentityMap because some fields
are null.

Awkan committed:
Fix default value of one-to-many order-by to ASC, #7141

JarJak committed:
Mention that Doctrine does not use Entities public API

JarJak committed:
Add info about Doctrine not using constructor

JarJak committed:
Unnecessary newline removed, small improvements in text

Majkl578 committed:
Add UPGRADE note for EntityRepository::count()

alexdenvir committed:
[XML] Fix default value of many-to-many order-by to ASC

SenseException committed:
Fix of links and anchors

asgrim committed:
$hydrationMode throughout can be a string as well as int (for custom modes)

Commit:
Fixed URLs of doctrine-mapping.xsd in docs

Until now the references to the `doctrine-mapping.xsd` consisted of different URLs.

A grep of docs showed:
* /Users/robo/dev/php/Doctrine/doctrine-mapping.xsd
* http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd
* http://www.doctrine-project.org/schemas/orm/doctrine-mapping.xsd
* https://raw.github.com/doctrine/doctrine2/master/doctrine-mapping.xsd

Now it is used http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd everywhere.

SenseException committed:
Fix of links and php version after review

lcobucci committed:
Use HTTPS endpoint for XML schema location

alcaeus committed:
Fix parameter value processing for objects with unloaded metadata

seferov committed:
fix incorrect phpdoc typehint

Majkl578 committed:
Lock dependencies for Code Quality stage

lcobucci committed:
Isolate entities used by the new test

To ensure we don't have any unintended side-effect.

Ocramius committed:
As previously reported by @flaushi in https://github.com/doctrine/doctrine2/pull/7471#discussion_r241949045, we discovered
that binding a parameter causes a `ClassMetadataFactory#getClassMetadata()` call, which in turn leads to large performance
regression when using any `object` type as parameter.

Following two snippets lead to an internal `ClassMetadataFactory#getClassMetadata()` call, which in turn leads to an
exception being thrown and garbage collected, plus multiple associated performance implications:

```php
$query->setParameter('foo', new DateTime());
$query->getResult();
```

```php
$query->setParameter('foo', new DateTime(), DateTimeType::NAME);
$query->getResult();
```

This is due to following portion of code:

https://github.com/doctrine/doctrine2/blob/434820973cadf2da2d66e7184be370084cc32ca8/lib/Doctrine/ORM/Query.php#L406-L409

Notice how `$value = $this->processParameterValue($value);` happens before attempting to infer the type for the parameter value.

That call leads to this segment being reached, which leads to the regression:

https://github.com/doctrine/doctrine2/blob/434820973cadf2da2d66e7184be370084cc32ca8/lib/Doctrine/ORM/AbstractQuery.php#L423-L433

Assuming the bound parameter type is provided, we can completely skip attempting to introspect the given object:

```php
$query->setParameter('foo', new DateTime(), DateTimeType::NAME);
$query->getResult();
```

Processing the parameter value is not needed in this case, so we can safely skip that logic for all known parameters.
In order to not introduce a BC break or change the `AbstractQuery#processParameterValue()` implementation, we could filter
out all parameters for which the type is given upfront, and later on merge them back in instead.

The test expectation to be set is for `UnitOfWork#getSingleIdentifierValue()` to never be called.

Ocramius committed:
Note: this will still lead to the `UnitOfWork#getSingleIdentifierValue()` still being
called when not specifying the type of a DQL parameter being bound via
`Doctrine\ORM\Query#setParameter()`:

```php
$query->setParameter('foo', $theValue, $theType);
```

A full parameter bind is required in order to gain back performance:

```php
$query->setParameter('foo', $theValue, $theType);
```

This is up for discussion with patch reviewers.

Ocramius committed:
As an example result:

```
./phpbench.phar run tests/Doctrine/Performance/Query --iterations=50 --revs=50 --report=aggregate
PhpBench 0.15-dev (dcbe193). Running benchmarks.
Using configuration file: /home/ocramius/Documents/doctrine/doctrine2/phpbench.json

\Doctrine\Performance\Query\QueryBoundParameterProcessingBench

    benchExecuteParsedQueryWithInferredParameterTypeI49 P0 	[μ Mo]/r: 643.684 634.664 (μs) 	[μSD μRSD]/r: 17.700μs 2.75%
    benchExecuteParsedQueryWithDeclaredParameterTypeI49 P0 	[μ Mo]/r: 97.673 94.251 (μs) 	[μSD μRSD]/r: 8.259μs 8.46%

2 subjects, 100 iterations, 100 revs, 0 rejects, 0 failures, 0 warnings
(best [mean mode] worst) = 88.460 [370.679 364.458] 127.400 (μs)
⅀T: 37,067.880μs μSD/r 12.980μs μRSD/r: 5.603%
suite: 133f0e30090f815142331ebec6af18241694e7c0, date: 2018-12-19, stime: 10:47:10
+------------------------------------+--------------------------------------------------+--------+--------+------+-----+------------+-----------+-----------+-----------+-----------+----------+--------+-------+
| benchmark                          | subject                                          | groups | params | revs | its | mem_peak   | best      | mean      | mode      | worst     | stdev    | rstdev | diff  |
+------------------------------------+--------------------------------------------------+--------+--------+------+-----+------------+-----------+-----------+-----------+-----------+----------+--------+-------+
| QueryBoundParameterProcessingBench | benchExecuteParsedQueryWithInferredParameterType |        | []     | 50   | 50  | 5,970,568b | 604.680μs | 643.684μs | 634.664μs | 677.640μs | 17.700μs | 2.75%  | 6.59x |
| QueryBoundParameterProcessingBench | benchExecuteParsedQueryWithDeclaredParameterType |        | []     | 50   | 50  | 5,922,424b | 88.460μs  | 97.673μs  | 94.251μs  | 127.400μs | 8.259μs  | 8.46%  | 1.00x |
+------------------------------------+--------------------------------------------------+--------+--------+------+-----+------------+-----------+-----------+-----------+-----------+----------+--------+-------+
```

This indicates that the performance impact for NOT declaring parameter types
explicitly is *MASSIVE*.

Ocramius committed:
#7527 automated CS checks

batwolf committed:
fix typo

`has to allow null values` vs `has to allows null values`

spirlici committed:
Update ordered-associations.rst

Fixed some typos

darrylhein committed:
change variable name

to make it consistent throughout document

lcobucci reviewed (changes requested):
@guilhermeblanco we're missing recent changes and we have some commits that aren't necessary:

46cd05f2f8e5864fbf3f1b1e8d041dcbe781e2fe and 9b04b0c1e9d7a008ca55cbb01c18a6aa801f5c08 come from the backported version https://github.com/doctrine/orm/pull/7069, which is not needed to be applied again since it was originally made against `master`.

The same seems to happen for 73eaa1941477bdf99257bc142432c51d7d3acf5e and 042804cc7bd48544c9c23233991c2dcb72a43765

772426d3da2baa8c8e32dc811e45774bba6c4b34 seems to be missing the [test case](https://github.com/doctrine/orm/pull/7291/files#diff-81252fa90b8e754dc458a8133ac94290).

Commits from https://github.com/doctrine/orm/pull/7471 and https://github.com/doctrine/orm/pull/7528 should be present here too (among other PRs).

Is it possible that you used an old reference of the branch `2.6` to do this sync?

guilhermeblanco commented:
@lcobucci merged+rebased again. Unit test suite now runs smoothly. =)

guilhermeblanco committed:
Swap properties via entity manager mock

guilhermeblanco committed:
Fix mock creation

lcobucci committed:
Remove useless nested switch case

This is simpler and doesn't make PHPCS go crazy when trying to indent
things properly.

lcobucci committed:
Remove duplicated tag from PHPUnit config

lcobucci reviewed (approved):
@guilhermeblanco great work, thanks!

I removed many commits that were already present in `master`, which were mostly incorrect. Now PHPCS and PHPBench are also passing :tada:
